### PR TITLE
Fix withdraw yields min rate

### PR DESCRIPTION
**Pull request by lazarmitic**

### CODE REVIEW

File: packages/tempus-client_v2/src/adapters/PoolDataAdapter.ts [hunk 1]
```diff
@@ -455,6 +456,15 @@ export default class PoolDataAdapter {
     }
 
     try {
+      let tokenSwapAmount = totalPrincipals;
+      if (totalYields.gt(totalPrincipals)) {
+        tokenSwapAmount = totalYields;
+      }
+
+      const estimatedYields = await this.getExpectedReturnForShareToken(tempusAMM, tokenSwapAmount, true);
```
Comment (danieledamiani) [inline packages/tempus-client_v2/src/adapters/PoolDataAdapter.ts:464]: what's `true`?

Comment (lazarmitic) [reply packages/tempus-client_v2/src/adapters/PoolDataAdapter.ts:464]: let's leave it like this for now I will fix it a bit later. We use TS so I always kinda assume we have all parameters named properly etc, so you can just jump to function definition and see the param name.

**Review by danieledamiani — DISMISSED** [review]

File: packages/tempus-client_v2/src/adapters/PoolDataAdapter.ts (hunk 1 above)
Comment (axic) [reply packages/tempus-client_v2/src/adapters/PoolDataAdapter.ts:464]: Shouldn't this be based on the above condition, whether tokenSwapAmount contains principals or yields?

Comment (lazarmitic) [reply packages/tempus-client_v2/src/adapters/PoolDataAdapter.ts:464]: Yup, I just realized that.

Comment (mijovic) [inline packages/tempus-client_v2/src/adapters/PoolDataAdapter.ts:464]: Just to document my suggestion from call 
```suggestion
      let tokenInAmount:BigNumber;
      let yieldsIn:bool;
      if (totalYields.gt(totalPrincipals)) {
        tokenInAmount = totalYields - totalPrincipals;
        yieldsIn = true;
      } else {
        tokenInAmount = totalPrincipals - totalYields;
        yieldsIn = false;
      }

      const estimatedTokensOut = await this.getExpectedReturnForShareToken(tempusAMM, tokenSwapAmount, yieldsIn);
```

File: packages/tempus-client_v2/src/adapters/PoolDataAdapter.ts [hunk 2]
```diff
@@ -455,6 +456,15 @@ export default class PoolDataAdapter {
...
+
+      const yieldsRate = div18f(estimatedYields, totalYields);
```
Comment (mijovic) [inline packages/tempus-client_v2/src/adapters/PoolDataAdapter.ts:466]:
```suggestion
      const yieldsRate = yieldsIn ? div18f(estimatedTokenOut, tokenInAmount) : div18f(tokenInAmount, estimatedTokenOut);
```

**Review by danieledamiani — APPROVED** [review]

File: packages/tempus-client_v2/src/adapters/PoolDataAdapter.ts [hunk 3]
```diff
@@ -455,6 +456,27 @@ export default class PoolDataAdapter {
...
+        tokenSwapAmount = totalPrincipals.sub(totalYields);
+
+        const estimatedYields = await this.getExpectedReturnForShareToken(tempusAMM, tokenSwapAmount, false);
+        yieldsRate = div18f(estimatedYields, tokenSwapAmount);
+      } else {
+        // In case we have equal amounts, use 1 as swapAmount just in case estimate was wrong, and swap is going to happen anyways
+        tokenSwapAmount = ethers.utils.parseEther('1');
+
+        const estimatedPrincipals = await this.getExpectedReturnForShareToken(tempusAMM, tokenSwapAmount, true);
+        yieldsRate = div18f(estimatedPrincipals, tokenSwapAmount);
```
Comment (mijovic) [inline packages/tempus-client_v2/src/adapters/PoolDataAdapter.ts:477]: Isn't it `yieldsRate = div18f(tokenSwapAmount, estimatedPrincipals);`

File: packages/tempus-client_v2/src/adapters/PoolDataAdapter.ts [hunk 4]
```diff
@@ -455,6 +456,27 @@ export default class PoolDataAdapter {
...
 
     try {
+      let yieldsRate;
+
+      let tokenSwapAmount;
+      if (totalYields.gt(totalPrincipals)) {
+        tokenSwapAmount = totalYields.sub(totalPrincipals);
+
+        const estimatedPrincipals = await this.getExpectedReturnForShareToken(tempusAMM, tokenSwapAmount, true);
+        yieldsRate = div18f(estimatedPrincipals, tokenSwapAmount);
```
Comment (mijovic) [inline packages/tempus-client_v2/src/adapters/PoolDataAdapter.ts:466]: Isnt it `yieldsRate = div18f(tokenSwapAmount, estimatedPrincipals);`

File: packages/tempus-client_v2/src/adapters/PoolDataAdapter.ts [hunk 5]
```diff
@@ -455,6 +456,27 @@ export default class PoolDataAdapter {
...
+      } else if (totalPrincipals.gt(totalYields)) {
+        tokenSwapAmount = totalPrincipals.sub(totalYields);
+
+        const estimatedYields = await this.getExpectedReturnForShareToken(tempusAMM, tokenSwapAmount, false);
+        yieldsRate = div18f(estimatedYields, tokenSwapAmount);
```
Comment (mijovic) [inline packages/tempus-client_v2/src/adapters/PoolDataAdapter.ts:471]: I think this should be `yieldsRate = div18f(tokenSwapAmount, estimatedYields);` as we are swaping principals for yields and yields rate is always `PrincipalsAmount / YieldsAmount`

Comment (mijovic) [reply packages/tempus-client_v2/src/adapters/PoolDataAdapter.ts:471]: we need to hold this property: `PrincipalsAmount = YieldsAmount * YieldsRate`

**Pull request merged**